src/OpenLoco/src/Drawing/SoftwareDrawingContext.cpp

```diff
@@ -1509,16 +1510,129 @@ namespace OpenLoco::Drawing
+            }
+
+            uint16_t crossPattern = 0;
+
+            int32_t leftX = left - rt.x;
+            if (leftX < 0)
+            {
+                crossPattern ^= leftX;
+                leftX = 0;
+            }
+
+            int32_t rightX = right - rt.x + 1;
+            if (rightX > rt.width)
+            {
+                rightX = rt.width;
+            }
+
+            int32_t topY = top - rt.y;
+            if (topY < 0)
+            {
+                crossPattern ^= topY;
+                topY = 0;
+            }
+
+            int32_t bottomY = bottom - rt.y + 1;
+            if (bottomY > rt.height)
+            {
+                bottomY = rt.height;
+            }
+
+            Rect drawRect = Rect::fromLTRB(leftX, topY, rightX, bottomY);
+
+            if (flags == RectFlags::none) // Regular fill
+            {
+                uint8_t* dst = drawRect.top() * (rt.width + rt.pitch) + drawRect.left() + rt.bits;
+                for (int32_t i = 0; i < drawRect.height(); i++)
+                {
+                    std::fill_n(dst, drawRect.width(), colour);
+                    dst += rt.width + rt.pitch;
+                }
+            }
+            else if ((flags & RectFlags::transparent) != RectFlags::none)
+            {
+                uint8_t* dst = rt.bits
+                    + static_cast<uint32_t>((drawRect.top() >> rt.zoomLevel) * ((rt.width >> rt.zoomLevel) + rt.pitch) + (drawRect.left() >> rt.zoomLevel));
+
+                // Find colour in colour table?
+                auto paletteMap = Gfx::PaletteMap::getForColour(static_cast<ExtColour>(colour));
+                if (paletteMap.has_value())
+                {
+                    const auto& paletteEntries = paletteMap.value();
+                    const int32_t scaled_width = drawRect.width() >> rt.zoomLevel;
+                    const int32_t step = (rt.width >> rt.zoomLevel) + rt.pitch;
+
+                    // Fill the rectangle with the colours from the colour table
+                    auto c = drawRect.height() >> rt.zoomLevel;
+                    for (int32_t i = 0; i < c; i++)
+                    {
+                        uint8_t* nextdst = dst + step * i;
+                        for (int32_t j = 0; j < scaled_width; j++)
+                        {
+                            auto index = *(nextdst + j);
+                            *(nextdst + j) = paletteEntries[index];
+                        }
+                    }
+                }
+            }
+            else if ((flags & RectFlags::crossHatching) != RectFlags::none)
+            {
+                uint8_t* dst = (drawRect.top() * (rt.width + rt.pitch)) + drawRect.left() + rt.bits;
+                for (int32_t i = 0; i < drawRect.height(); i++)
+                {
+                    uint8_t* nextdst = dst + rt.width + rt.pitch;
+                    uint32_t p = Numerics::ror32(crossPattern, 1);
+                    p = (p & 0xFFFF0000) | drawRect.width();
```

Comment:
This is pretty hard to understand due to CS using the same registers for two things. It would be simpler to do the following:

```cpp
                    uint32_t p = Numerics::ror32(crossPattern, 1);
                    for (auto x = 0; x < drawRect.width(); x++)
                    {
                         p ^= 0x80000000;
                         if (p & 0x80000000){
                             *dst = colour;
                         }
                         dst++;
                     }
                    crossPattern ^= 1;
                    dst = nextdst;
```
Could probably actually copy more of the design of transparent with respect to the nextDst, step and x, y. That would make it even easier to follow and show the similarities (and the curious fact that transparency works on zoomed rt's)